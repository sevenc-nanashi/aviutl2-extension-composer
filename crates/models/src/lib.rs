mod generated {
    include!(env!("SCHEMA_RS_PATH"));
}
pub use generated::*;
pub use overrides::*;

mod overrides {
    #[allow(non_camel_case_types)]
    pub type u64 = ::std::primitive::u64;

    pub type Url = ::url::Url;

    pub mod license {
        macro_rules! create_license_struct {
            ($name:ident, $str_val:expr) => {
                #[derive(Debug, Clone, PartialEq, Eq)]
                pub struct $name;

                impl serde::Serialize for $name {
                    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
                    where
                        S: serde::Serializer,
                    {
                        serializer.serialize_str($str_val)
                    }
                }

                impl<'de> serde::Deserialize<'de> for $name {
                    fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
                    where
                        D: serde::Deserializer<'de>,
                    {
                        let ss = String::deserialize(deserializer)?;
                        if ss == $str_val {
                            Ok($name)
                        } else {
                            Err(<D::Error as serde::de::Error>::invalid_value(
                                serde::de::Unexpected::Str(&ss),
                                &concat!("`", $str_val, "`"),
                            ))
                        }
                    }
                }
            };
        }
        create_license_struct!(Free, "free");
        create_license_struct!(Nicovideo, "nicovideo");
        create_license_struct!(Custom, "custom");
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_parsing_example_manifest() {
        insta::assert_debug_snapshot!(
            serde_yml::from_str::<Manifest>(include_str!("../../../docs/examples/manifest.yml"))
                .unwrap()
        );
    }

    #[test]
    fn test_parsing_example_registry() {
        insta::assert_debug_snapshot!(
            serde_yml::from_str::<Registry>(include_str!("../../../docs/examples/registry.yml"))
                .unwrap()
        );
    }
}
