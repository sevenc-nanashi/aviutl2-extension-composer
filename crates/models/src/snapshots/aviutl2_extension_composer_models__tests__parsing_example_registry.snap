---
source: crates/models/src/lib.rs
expression: "serde_yml::from_str::<Registry>(include_str!(\"../../../docs/examples/registry.yml\")).unwrap()"
---
Registry {
    contents: [
        RegistryContentsItem {
            authors: [
                RegistryContentsItemAuthorsItem {
                    name: Single(
                        "Nanashi.",
                    ),
                    url: Some(
                        HttpUrl(
                            "https://sevenc7c.com",
                        ),
                    ),
                },
            ],
            description: Some(
                Single(
                    "Rustで書かれたFFmpeg出力プラグインです。\n...\n（省略）\n...",
                ),
            ),
            homepage: Some(
                "https://github.com/sevenc-nanashi/aviutl2-rs",
            ),
            id: ManifestId(
                "sevenc_nanashi-aviutl2_rs-ffmpeg_output",
            ),
            license: Nicovideo {
                id: "sm45355531",
                name: Nicovideo,
                omittable: Some(
                    true,
                ),
                text: None,
            },
            manifest_url: "https://sevenc7c.example.com/ffmpeg_output.manifest.yml",
            manifest_version: 1,
            name: "Rusty FFmpeg Output",
            summary: Single(
                "Rustで書かれたFFmpeg出力プラグイン",
            ),
            tags: [
                Single(
                    "#plugin-output",
                ),
            ],
            version: "0.1.0",
            version_number: Some(
                1,
            ),
        },
    ],
    homepage: Some(
        "https://github.com/sevenc-nanashi/aviutl2-rs",
    ),
    name: "AviUtl2-rs Plugin Registry",
    registry_version: 1.0,
}
